### ISSUE TITLE
Do not import all pools autmatically inside zfs-utils.initcpio.hook

### ISSUE THREAD

**Issue by Bronek** [archzfs/archzfs#50]
Automatically importing all pools inside `zfs-utils.initcpio.hook` causes problems, for example loss of disk names inside `zpool.cache`, pool corruption on multipath SAN systems where the same pools might be already used by a different system at the same time, or slow boot times (and possibly pool corruption) if pools imported contain ZVOLs which themselves are ZFS pools (e.g. setup for virtual machines). This has been extensively discussed in ZOL at https://github.com/zfsonlinux/zfs/pull/3800 and https://github.com/zfsonlinux/zfs/issues/3777 and the agreement (with some reservations) is that ZOL should import pools from `zpool.cache` only, rather than import all pools. Archzfs already correctly installs systemd service `zfs-import-cache.service` which performs this without problems. Thus `zpool import -aN` should be removed from `zfs-utils.initcpio.hook` . This is what patch below does

```
diff --git a/zfs-utils.initcpio.hook b/zfs-utils.initcpio.hook
index 00948d4..80f8a2a 100644
--- a/zfs-utils.initcpio.hook
+++ b/zfs-utils.initcpio.hook
@@ -93,8 +93,4 @@ run_hook() {
 }


-run_latehook () {
-    /usr/bin/zpool import -N -a $ZPOOL_FORCE
-}
-
```

Obviously md5sum of `zfs-utils.initcpio.hook` should be updated inside PKGBUILD as appropriate

**Comment by techmunk** [comment]
I was responsible for adding the run_latehook() in #25. Whilst I get the points referenced in the other issues, just removing this, would not solve my use-case. Would adding the -c /etc/zfs/zpool.cache instead of removing this fix both our issues?

The other option, is to have some way to define in the initcpio, what ZFS datasets need to be mounted before the pivot root, so that they are available before the system boots. /var is the only real special case that I'm aware of, as if it is mounted too late, it will refuse to mount due to files getting written to it by other processes before the systemd zfs-import-cache.service runs.

Perhaps the cache service needs to run sooner?

**Comment by Bronek** [comment]
Yes if we added `-c /etc/zfs/zpool.cache` then cache will be used, in which case zfs-import-cache is superfluous and will just confuse user, and IMO should be removed OR alternatively ZOL should update zfs.target to start it earlier (my preferred option)

**Comment by demizer** [comment]
I am leaning towards @Bronek's recommendation here. We should be using the provided zfs-import-cache.service. For reference, here is the contents of the service file:

```
[Unit]
Description=Import ZFS pools by cache file
DefaultDependencies=no
Requires=systemd-udev-settle.service
After=systemd-udev-settle.service
After=cryptsetup.target
ConditionPathExists=/etc/zfs/zpool.cache

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStartPre=/sbin/modprobe zfs
ExecStart=/usr/bin/zpool import -c /etc/zfs/zpool.cache -aN
```

I will need to test this of course so @techmunk's use case in not broken. If it is, then we will need to raise an issue upstream, if they have not solved this already.

**Comment by kerberizer** [comment]
I'm somewhat worried that deviating significantly from upstream is likely to increase our unique use cases and problems that we would then have to solve on our own. Even if we use zpool.cache, importing _all_ pools in initramfs might still be unexpected for some people and cause unforeseen consequences. I'd rather vote for a patch to fix @techmunk's _specific_ problem, without affecting anyone else. The easiest seems to be a new kernel parameter with ZFS pools to be explicitly imported in initramfs, even if they don't contain a root dataset. People like techmunk might then use this parameter to suit their needs, while all the rest will use the unchanged setup from upstream, which is (likely) shared with most other distros.

**Comment by Bronek** [comment]
ZOL is about to change the behaviour of zfs-import*.service in release 0.6.5.8 , for details see https://github.com/zfsonlinux/zfs/pull/4699 . I asked whether that would solve #25

**Comment by storrgie** [comment]
I might not be asking in the right area, but I'm seeing my pools require a `zpool import` after I boot. I'm using luks/dm-crypt with keyfiles, so the devices don't show up in /dev/mapper until after that all gets executed.

Should I be editing a systemd unit file to postpone what would be natural at boot time?

**Comment by demizer** [comment]
systemd handling of pool imports is changed for 0.6.5.8 see #72

**Comment by demizer** [comment]
If anyone want's to take up the cause, this would be a good candidate for an automated test. We could copy https://github.com/archzfs/archzfs/tree/master/testing/tests/archzfs-qemu-lts-test-00-bootfs and add luks encryption. Be warned though, no one other than myself has setup an automated test for archzfs. There might be a yak to shave.

**Comment by v1ncente** [comment]
I'm a ZFS newbie and i don't even use Arch at the moment (but plan to in the near future) and i may be wrong, but why don't you use the ZFS_POOL_IMPORT variable from the ZoL userland configuration file
(https://github.com/zfsonlinux/zfs/blob/master/etc/init.d/zfs.in) to specify which pools to import in zfs-utils.initcpio.hook ? It seems to be made for this situation.
If this isn't desired i second a 'new kernel parameter' for techmunk's special case.

**Comment by samis** [comment]
This causes an annoying pause in boot for me, because it tries and fails to import pools for my illumos VMs which are stored in zvols on one of the host's pools.

**Comment by kcwitt** [comment]
I wrote a script called `paczfs` at https://github.com/varasys/paczfs which creates a bootable ArchLinux image using ZFS as the root filesystem (based on this repository).

In our use case, we ended up setting the mountpoint for datasets that can be used as the "root filesystem" to "legacy" and all other datasets use normal zfs mountpoints.

We also used the "bootfs=pool/dataset" kernel arguments (instead of the root=ZFS=pool/dataset method).

We had to edit the mkinitcpio zfs hook to "short circuit" checking the initramfs /etc/fstab file for an entry for the root filesystem before mounting it using the following:
```
sed -i '/"legacy")/a \                mount -t zfs "${dataset}" "${node}\/"; continue' "${CHROOT_PATH}/usr/lib/initcpio/hooks/zfs"
```

So under our use case, the zfs mkinitcpio script imports all pools, then mounts whichever dataset the pool's "bootfs" property points to, then continues. All other zfs datasets are mounted by the zfs-mount.service during the normal startup process (but in our use case /var /usr and /etc are all on the same rootfs so they are all mounted at the same time).

Just a reference "use-case" to be considered in any changes to the zfs hook script.

Also, we don't really understand zfs cache files and how they are used. We set the "cachefile=none" property for all our pools and haven't seen any issues .
